File: backend/Kernel/Canonical.cs
using System.Globalization;
using System.Text;

namespace Kernel;

/// <summary>
/// Anhang D — Wert eines Ketteneintrags.
///
/// Absichtlich ein eigener Typ und nicht System.Text.Json: 22.3 verbietet
/// Gleitkommazahlen, und ein Typ, der sie gar nicht erst aufnehmen kann, ist die
/// verlaesslichere Durchsetzung als eine Pruefung beim Serialisieren.
/// </summary>
public abstract record CanonicalJson
{
    public sealed record Str(string Value) : CanonicalJson;
    public sealed record Int(long Value) : CanonicalJson;
    public sealed record Bool(bool Value) : CanonicalJson;
    public sealed record Null : CanonicalJson;
    public sealed record Arr(IReadOnlyList<CanonicalJson> Items) : CanonicalJson;
    public sealed record Obj(IReadOnlyDictionary<string, CanonicalJson> Fields) : CanonicalJson;

    public static CanonicalJson S(string v) => new Str(v);
    public static CanonicalJson I(long v) => new Int(v);
    public static CanonicalJson B(bool v) => new Bool(v);
    public static CanonicalJson Nil => new Null();
    public static CanonicalJson G(Guid v) => new Str(Ids.ToText(v));
    public static CanonicalJson Hex(ReadOnlySpan<byte> v) => new Str(Crypto.ToHex(v));

    /// <summary>UTC, Sekundengenauigkeit, Zulu-Form. Ohne feste Form waeren zwei
    /// gleichwertige Zeitpunkte zwei verschiedene Hashes.</summary>
    public static CanonicalJson T(DateTimeOffset v) =>
        new Str(v.ToUniversalTime().ToString("yyyy-MM-dd'T'HH:mm:ss'Z'", CultureInfo.InvariantCulture));

    public static CanonicalJson A(params CanonicalJson[] items) => new Arr(items);
    public static CanonicalJson O(params (string Key, CanonicalJson Value)[] fields)
        => new Obj(fields.ToDictionary(f => f.Key, f => f.Value));
}

/// <summary>
/// Anhang D — Kanonische Serialisierung nach RFC 8785 (JCS), beschraenkt auf
/// Ganzzahlen.
///
/// Dies ist die haerteste Einbahntuer des Vorhabens: Sobald der erste Eintrag
/// geschrieben ist, macht jede Aenderung hier JEDE bestehende Signatur ungueltig.
/// Eine Aenderung an dieser Datei ist niemals eine Verbesserung, sondern immer
/// eine neue Kette.
/// </summary>
public static class Canonical
{
    /// <summary>2^53-1 — darueber wird als Zeichenkette gefuehrt (22.3).</summary>
    public const long MaxSafeInteger = 9007199254740991L;

    public static string Serialize(CanonicalJson value)
    {
        var sb = new StringBuilder(512);
        Write(sb, value);
        return sb.ToString();
    }

    /// <summary>UTF-8, ohne Byte Order Mark (22.2).</summary>
    public static byte[] SerializeToUtf8(CanonicalJson value)
        => new UTF8Encoding(encoderShouldEmitUTF8Identifier: false).GetBytes(Serialize(value));

    public static byte[] Hash(CanonicalJson value)
        => System.Security.Cryptography.SHA256.HashData(SerializeToUtf8(value));

    private static void Write(StringBuilder sb, CanonicalJson v)
    {
        switch (v)
        {
            case CanonicalJson.Null:
                sb.Append("null");
                break;

            case CanonicalJson.Bool b:
                sb.Append(b.Value ? "true" : "false");
                break;

            case CanonicalJson.Int i:
                if (i.Value > MaxSafeInteger || i.Value < -MaxSafeInteger)
                    throw new InvalidOperationException(
                        $"Ganzzahl {i.Value} liegt ausserhalb des sicheren Bereichs. Groessere Zahlen sind Zeichenketten (22.3).");
                sb.Append(i.Value.ToString(CultureInfo.InvariantCulture));
                break;

            case CanonicalJson.Str s:
                WriteString(sb, s.Value);
                break;

            case CanonicalJson.Arr a:
                sb.Append('[');
                for (var k = 0; k < a.Items.Count; k++)
                {
                    if (k > 0) sb.Append(',');
                    Write(sb, a.Items[k]);
                }
                sb.Append(']');
                break;

            case CanonicalJson.Obj o:
                sb.Append('{');
                // 22.2: aufsteigend nach UTF-16-Codeeinheiten. CompareOrdinal
                // vergleicht genau so — nicht nach UTF-8-Bytes und nicht nach
                // Sprachregeln.
                var keys = o.Fields.Keys.ToArray();
                Array.Sort(keys, string.CompareOrdinal);
                for (var k = 0; k < keys.Length; k++)
                {
                    if (k > 0) sb.Append(',');
                    WriteString(sb, keys[k]);
                    sb.Append(':');
                    Write(sb, o.Fields[keys[k]]);
                }
                sb.Append('}');
                break;

            default:
                throw new InvalidOperationException($"Unbekannter Werttyp {v.GetType().Name}.");
        }
    }

    /// <summary>
    /// 22.2 — Minimale Maskierung: nur " \ und Steuerzeichen. Nicht-ASCII bleibt
    /// unmaskiert und steht als UTF-8. Das ist der haeufigste Umsetzungsfehler
    /// (siehe TV-8).
    /// </summary>
    private static void WriteString(StringBuilder sb, string s)
    {
        sb.Append('"');
        foreach (var c in s)
        {
            switch (c)
            {
                case '"': sb.Append("\\\""); break;
                case '\\': sb.Append("\\\\"); break;
                case '\b': sb.Append("\\b"); break;
                case '\f': sb.Append("\\f"); break;
                case '\n': sb.Append("\\n"); break;
                case '\r': sb.Append("\\r"); break;
                case '\t': sb.Append("\\t"); break;
                default:
                    if (c < 0x20)
                        sb.Append("\\u").Append(((int)c).ToString("x4", CultureInfo.InvariantCulture));
                    else
                        sb.Append(c);
                    break;
            }
        }
        sb.Append('"');
    }
}

File: frontend/src/app/canonical.ts
/**
 * Anhang D — Kanonische Form, Browserseite.
 *
 * Sie MUSS Zeichen für Zeichen dasselbe liefern wie `Kernel/Canonical.cs`.
 * Über genau diesen Bytes wird unterschrieben: weicht eine Klammer, eine
 * Reihenfolge oder eine Maskierung ab, ist die Unterschrift ungültig — und der
 * Fehler zeigt sich nicht hier, sondern als „Unterschrift passt nicht" an einer
 * Stelle, die nichts dafür kann.
 *
 * <b>Die härteste Einbahntür.</b> Sobald die erste Kante unterschrieben ist,
 * macht jede Änderung an dieser Datei ALLE bestehenden Unterschriften ungültig.
 * Eine Änderung hier ist nie eine Verbesserung, sondern immer eine neue Kette.
 *
 * Keine Gleitkommazahlen: 22.3 verbietet sie, und ein Typ, der sie gar nicht
 * erst aufnimmt, ist die verlässlichere Durchsetzung als eine Prüfung beim
 * Schreiben.
 */

export type Canon =
  | { readonly k: 'str'; readonly v: string }
  | { readonly k: 'int'; readonly v: number }
  | { readonly k: 'bool'; readonly v: boolean }
  | { readonly k: 'null' }
  | { readonly k: 'arr'; readonly v: readonly Canon[] }
  | { readonly k: 'obj'; readonly v: Readonly<Record<string, Canon>> };

export const S = (v: string): Canon => ({ k: 'str', v });
export const B = (v: boolean): Canon => ({ k: 'bool', v });
export const NIL: Canon = { k: 'null' };
export const A = (...items: readonly Canon[]): Canon => ({ k: 'arr', v: items });
export const O = (fields: Readonly<Record<string, Canon>>): Canon => ({ k: 'obj', v: fields });

/** 2^53-1 — darüber führt 22.3 Zahlen als Zeichenketten. */
export const MAX_SAFE_INTEGER = 9007199254740991;

export function I(v: number): Canon {
  if (!Number.isInteger(v)) throw new Error(`Nur Ganzzahlen: ${v} ist keine.`);
  if (v > MAX_SAFE_INTEGER || v < -MAX_SAFE_INTEGER) {
    throw new Error(`Ganzzahl ${v} liegt ausserhalb des sicheren Bereichs (22.3).`);
  }
  return { k: 'int', v };
}

/**
 * 22.2 — Minimale Maskierung: nur `"`, `\` und Steuerzeichen. Nicht-ASCII
 * bleibt unmaskiert und steht als UTF-8. Das ist der häufigste
 * Umsetzungsfehler, und er fällt erst bei einem Namen mit Umlaut auf.
 */
function writeString(out: string[], s: string): void {
  out.push('"');
  for (const c of s) {
    switch (c) {
      case '"': out.push('\\"'); break;
      case '\\': out.push('\\\\'); break;
      case '\b': out.push('\\b'); break;
      case '\f': out.push('\\f'); break;
      case '\n': out.push('\\n'); break;
      case '\r': out.push('\\r'); break;
      case '\t': out.push('\\t'); break;
      default: {
        const code = c.codePointAt(0) ?? 0;
        out.push(code < 0x20 ? `\\u${code.toString(16).padStart(4, '0')}` : c);
      }
    }
  }
  out.push('"');
}

function write(out: string[], v: Canon): void {
  switch (v.k) {
    case 'null': out.push('null'); break;
    case 'bool': out.push(v.v ? 'true' : 'false'); break;
    case 'int': out.push(String(v.v)); break;
    case 'str': writeString(out, v.v); break;

    case 'arr':
      out.push('[');
      v.v.forEach((item, i) => { if (i > 0) out.push(','); write(out, item); });
      out.push(']');
      break;

    case 'obj': {
      out.push('{');

      /*
       * 22.2: aufsteigend nach UTF-16-Codeeinheiten. Genau das tut der
       * Standardvergleich von `sort` — und genau das tut `string.CompareOrdinal`
       * auf der anderen Seite. Ein `localeCompare` wäre hier der Fehler, den man
       * erst bei einem Feldnamen mit Umlaut bemerkt.
       */
      const keys = Object.keys(v.v).sort();
      keys.forEach((key, i) => {
        if (i > 0) out.push(',');
        writeString(out, key);
        out.push(':');
        write(out, v.v[key]);
      });

      out.push('}');
      break;
    }
  }
}

export function serialize(value: Canon): string {
  const out: string[] = [];
  write(out, value);
  return out.join('');
}

/** UTF-8, ohne Byte Order Mark (22.2). */
export const toUtf8 = (value: Canon): Uint8Array => new TextEncoder().encode(serialize(value));
